torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1059,6 +1058,41 @@ void ProcessGroupNCCL::eagerConnectSingleDevice(at::Device device) {
   getNCCLComm(key, device, OpType::ALLREDUCE);
 }
 
+bool ProcessGroupNCCL::useNonblocking() {
+#ifndef NCCL_HAS_COMM_NONBLOCKING
+  return false;
+#endif
+  // Already parsed, return the cached value
+  if (useNonblocking_.has_value()) {
+    return useNonblocking_.value();
+  }
+  // Get environment variable.
+  auto nbEnv = c10::utils::check_env("TORCH_NCCL_USE_COMM_NONBLOCKING");
+
+  // 1st priority: Respect the user's setting
+  if (options_->config.blocking != NCCL_CONFIG_UNDEF_INT) {
+    useNonblocking_ = options_->config.blocking == 0;
+    goto print_and_return;
+  }
+  // 2nd priority: Respect the environment variable
+  if (nbEnv.has_value()) {
+    useNonblocking_ = nbEnv.value();
+    goto print_and_return;
+  }
+  // 3rd priority: automatically use nonblocking if we are in eager init mode
+  if (getBoundDeviceId()) {
+    useNonblocking_ = true;
+    goto print_and_return;
+  }
+  // 4th priority: otherwise, nonblocking = false to preserve old behavior
+  useNonblocking_ = false;
+
+print_and_return:
+  LOG(INFO) << logPrefix()
```

Comment:
nit: this is ok, but i think its also useful to log which of the config pathways took precedent, in case users want to debug why their nonblocking setting wasn't as expected